File: backend/src/sqs/handler.js
import { connectDB } from '../database/db.js';
import { routeIncomingAiJob } from './router.js';


export const handler = async (event, context) => {
  // Ensure connection instance is running
  await connectDB();
  
  const batchItemFailures = [];

  if (event && Array.isArray(event.Records)) {
    for (const record of event.Records) {
      try {
        const message = JSON.parse(record.body);
        await routeIncomingAiJob(message);
      } catch (error) {
        console.error(
          `[SQS Handler Critical] Failed to completely digest message ID ${record.messageId}: ${error.message}`
        );
        // Track the specific failed item ID so SQS can retry it individually
        batchItemFailures.push({ itemIdentifier: record.messageId });
      }
    }
  }

  return { batchItemFailures };
};

File: backend/app/sqs/handler.py
# app/sqs/handler.py

import asyncio
import json
from app.database import init_db, close_db
from .router import route_incoming_ai_job  

async def start_async_queue_handler(event, context):

    await init_db()
    batch_item_failures = []
    
    try:
        for record in event.get("Records", []):
            try:
                message = json.loads(record["body"])
                await route_incoming_ai_job(message)
            except Exception as e:
                message_id = record.get("messageId")
                print(f"[SQS Handler Critical] Failed to completely digest message ID {message_id}: {str(e)}")
                if message_id:
                    batch_item_failures.append({"itemIdentifier": message_id})
                
        return {"batchItemFailures": batch_item_failures}
    finally:
        await close_db()


def handler(event, context):
    return asyncio.run(start_async_queue_handler(event, context))